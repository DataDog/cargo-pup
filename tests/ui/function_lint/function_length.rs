//@compile-flags: --crate-name test_function_lint
//@compile-flags: --crate-type lib

// This test verifies that the FunctionLint's MaxLength rule works correctly

// ====== Basic Function Length Tests ======

// Function exceeds the 5-line limit (rule: function_length_test)
fn too_long_function() { //~ ERROR: Function exceeds maximum length of 5 lines with 6 lines
    let a = 1;
    let b = 2;
    let c = 3;
    println!("This function has too many lines");
}

// This function is within the limit and should not trigger an error
fn acceptable_length() {
    let a = 1;
    let b = 2;
    let c = 3;
}

// ====== Name-based Function Length Tests ======

// This function matches the name-based rule (3 lines max for functions with "another_long" prefix)
fn another_long_function() { //~ ERROR: Function exceeds maximum length of 3 lines with 5 lines
    let x = 1;
    let y = 2;
    println!("This is also too long");
}

// This function doesn't match the name pattern and won't trigger an error
fn long_but_ok() {
    println!("This is OK");
}

// ====== Methods in Structs ======

pub struct TestStruct;

impl TestStruct {
    // This method is too long for the general 5-line rule
    fn too_long_method(&self) { //~ ERROR: Function exceeds maximum length of 5 lines with 6 lines
        let a = 1;
        let b = 2;
        let c = 3;
        println!("This method has too many lines");
    }
    
    // This method matches the name-based rule and exceeds the 3-line limit
    fn another_long_method(&self) { //~ ERROR: Function exceeds maximum length of 3 lines with 5 lines
        let x = 1;
        let y = 2;
        println!("This is also too long");
    }
    
    // This method is OK (3 lines total)
    fn short_method(&self) {
        println!("This method is OK");
    }
}

// ====== Module-based Function Length Tests ======

mod inner_module {
    // Even a tiny function exceeds the 2-line limit for this module
    pub fn tiny_but_too_long() { //~ ERROR: Function exceeds maximum length of 2 lines with 3 lines
        println!("Just one line makes this too long");
    }
    
    // This function is OK (2 lines total)
    pub fn just_right() {}
    
    // A nested module that should also have the 2-line limit
    pub mod nested {
        // This exceeds the 2-line limit inherited from the parent module path rule
        pub fn nested_too_long() { //~ ERROR: Function exceeds maximum length of 2 lines with 3 lines
            println!("This violates the 2-line limit");
        }
        
        // This is OK (2 lines)
        pub fn nested_fine() {}
    }
} 