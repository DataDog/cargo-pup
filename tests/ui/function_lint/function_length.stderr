error: Function exceeds maximum length of 5 lines with 6 lines
  --> tests/ui/function_lint/function_length.rs:11:1
   |
LL | fn too_long_function() {
   | ^^^^^^^^^^^^^^^^^^^^^^
   |
   = help: Consider breaking this function into smaller parts
   = note: Applied by cargo-pup rule 'function_length_test'.
   = note: `#[deny(function_lint)]` on by default

error: Function exceeds maximum length of 3 lines with 5 lines
  --> tests/ui/function_lint/function_length.rs:28:1
   |
LL | fn another_long_function() {
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = help: Consider breaking this function into smaller parts
   = note: Applied by cargo-pup rule 'named_function_length_test'.

error: Function exceeds maximum length of 5 lines with 6 lines
  --> tests/ui/function_lint/function_length.rs:45:5
   |
LL |     fn too_long_method(&self) {
   |     ^^^^^^^^^^^^^^^^^^^^
   |
   = help: Consider breaking this function into smaller parts
   = note: Applied by cargo-pup rule 'function_length_test'.

error: Function exceeds maximum length of 3 lines with 5 lines
  --> tests/ui/function_lint/function_length.rs:53:5
   |
LL |     fn another_long_method(&self) {
   |     ^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = help: Consider breaking this function into smaller parts
   = note: Applied by cargo-pup rule 'named_function_length_test'.

error: Function exceeds maximum length of 2 lines with 3 lines
  --> tests/ui/function_lint/function_length.rs:69:5
   |
LL |     pub fn tiny_but_too_long() {
   |     ^^^^^^^^^^^^^^^^^^^^^^
   |
   = help: Consider breaking this function into smaller parts
   = note: Applied by cargo-pup rule 'module_based_function_length_test'.

error: Function exceeds maximum length of 2 lines with 3 lines
  --> tests/ui/function_lint/function_length.rs:79:9
   |
LL |         pub fn nested_too_long() {
   |         ^^^^^^^^^^^^^^^^^^^^
   |
   = help: Consider breaking this function into smaller parts
   = note: Applied by cargo-pup rule 'module_based_function_length_test'.

error: aborting due to 6 previous errors

